Push tracing of Parlia system transactions so that live tracers can properly traces those state changes

maoueh opened the pull request:
### Description

See #2768 for context and details.

This introduced `OnSystemTxStart/OnSystemTxEnd` and renamed `OnSystemTxEnd` to `OnSystemTxFixIntrinsicGas` (so `OnSystemTxEnd` can be re-used for other purposes).

The Parlia engine is now tracing it's system transaction properly which was not the case before.

### Rationale

This PR ensure that system transactions are properly tested when using live tracers.

### Changes

Consensus engine is now receiving a `tracer *tracing.Hooks` that Parlia engine then uses when needing to apply system transactions.

maoueh commented on consensus/parlia/parlia.go:
```diff
@@ -1949,23 +1957,38 @@ func (p *Parlia) applyTransaction(
+	txs *[]*types.Transaction, receipts *[]*types.Receipt,
+	receivedTxs *[]*types.Transaction, usedGas *uint64, mining bool,
+	vmConfig vm.Config,
+) (receipt *types.Receipt, applyErr error) {
```
It's for the `defer` to pass it down to `OnTxEnd(receipt)` at https://github.com/bnb-chain/bsc/pull/2772/files/a69186357e1be7043932f26dd621cadb22342c13#diff-7c199a2e5208a5ff8460e89bc19a9f9734312a8a4b15052d9d3e7e65c74aaa06R2022.

I initially modified `applyTransaction` but felt it was bad to change all call site to right away ignore so I created a second version.

But I could implemented differently maybe that would be less confusing:

```go
func applyTransaction(...) error {
  // Used by OnTxEnd via a defer to properly pass the final receipt.
  var tracingReceipt *types.Receipt
...
```

I usually prefer to use Go native return capabilities which works even if `receipt` is not set explicitly but only `return`.

Let me know the team preference here, we can also keep the current code and I can document the function to explain `receipt` usage.

zzzckck replied:
@maoueh I would prefer the second one: create the receipt variable at the beginning
``` golang
func applyTransaction(...) error {
  // Used by OnTxEnd via a defer to properly pass the final receipt.
  var tracingReceipt *types.Receipt
...

zzzckck commented on consensus/parlia/parlia.go:
```diff
@@ -1840,15 +1843,20 @@ func (p *Parlia) distributeIncoming(val common.Address, state *state.StateDB, he
-	state.SetBalance(consensus.SystemAddress, common.U2560, tracing.BalanceChangeUnspecified)
-	state.AddBalance(coinbase, balance, tracing.BalanceChangeUnspecified)
+
+	// FIXME Those need some balance change reason, in our current Firehose implementation we had put
+	// firehose.BalanceChangeReason("reward_transfaction_fee"), this would map to
+	// `tracing.BalanceIncreaseRewardTransactionFee` in the new tracing API, is that accurate enough here?
+	// I would be fine adding a new more specific reason if needed.
+	state.SetBalance(consensus.SystemAddress, common.U2560, tracing.BalanceIncreaseRewardTransactionFee)
+	state.AddBalance(coinbase, balance, tracing.BalanceIncreaseRewardTransactionFee)
```
@maoueh I think `BalanceIncreaseRewardTransactionFee` is not accurate. I would prefer to introduce the two new reason for BSC, but I am afraid of forward compatible issue, how about:
```
BalanceDecreaseBSCDistributeReward BalanceChangeReason = 1000
BalanceIncreaseBSCDistributeReward BalanceChangeReason = 1001
```

maoueh replied:
I prefer indeed to add 2 new reasons, makes tracing system aware of important balance changes. Don't think Unspecified is good here.

> but I am afraid of forward compatible issue

Indeed, that's always a possibility, even more that the `BalanceChangeReason` is a `byte` today leading to just 255 values.

But having managed multiple Geth forks over the last 3 years, I can tell you that new balance change reasons is very rare. I think it will be just fine. I've put BSC starting at 210 for now, let me know how it feels.

Joj501 reviewed (approved):
Approve change

maoueh commented:
@zzzckck @Joj501 I adjusted the PR

galaio commented on consensus/parlia/parlia.go:
```diff
@@ -1979,7 +1985,36 @@ func (p *Parlia) applyTransaction(
+	context := core.NewEVMBlockContext(header, chainContext, nil)
+	// Create a new environment which holds all relevant information
+	// about the transaction and calling mechanisms.
+	vmenv := vm.NewEVM(context, vm.TxContext{Origin: msg.From(), GasPrice: big.NewInt(0)}, state, p.chainConfig, vm.Config{})
```
Here is missing the tracer, may
```go
vmenv := vm.NewEVM(context, vm.TxContext{Origin: msg.From(), GasPrice: big.NewInt(0)}, state, p.chainConfig, vmConfig)
```

maoueh replied:
Yeah good catch, how could I miss that, that was the whole point :$

galaio commented on miner/worker.go:
```diff
@@ -1166,7 +1166,7 @@ func (w *worker) generateWork(params *generateParams, witness bool) *newPayloadR
 	body := types.Body{Transactions: work.txs, Withdrawals: params.withdrawals}
 	fees := work.state.GetBalance(consensus.SystemAddress)
-	block, receipts, err := w.engine.FinalizeAndAssemble(w.chain, work.header, work.state, &body, work.receipts)
+	block, receipts, err := w.engine.FinalizeAndAssemble(w.chain, work.header, work.state, &body, work.receipts, nil)
```
Why not add a tracer here? Same question on line 1432.

maoueh replied:
Miner should not trace block as their are being produced so they cannot be relied on. Their could be a block produced by the miner and a block being processed in state processor at the same time. 

The Geth 1.14.x behaves the same for live tracer(s), they are not invoked by `miner` or optimistic processing.

galaio commented on core/tracing/hooks.go:
```diff
@@ -150,9 +150,41 @@ type (
+	// This event flow enables transactions to be traced correctly with just OnTxStart/End being set but also
+	// enables special routing of those transactions by having OnSystemTxStart/End defined and keeping a special
+	// system state to do something different when OnTxEnd is called.
+	OnSystemTxStartHook func()
```
It's not necessary to add `OnSystemTxStartHook` or `OnSystemTxEndHook`, it can be merged into `OnTxStart` and `OnTxEnd`, and it is more flexible too.

```go
//TxStartHook 
func(vm *VMContext, tx *types.Transaction, from common.Address) {
  // check if system contract first
  if tx.To() != nil && from == vm.Coinbase && parlia.isToSystemContract(*tx.To()) && tx.GasPrice().Cmp(big.NewInt(0)) == 0 {
      //handle system tx
   }
}
```

maoueh replied:
`parlia` wouldn't be accessible from tracer. I agree that if within the tracer when in `OnTxStart/End` there would be a way to determine if the transaction is a system one, I think it would be ok.

However, I still prefer an explicit hooks:
- If someone does not implement `OnSystemTxStartHook/End`, everything still works.
- It makes it a bit more explicit for someone to look into the tracer Hooks and understand there could be system transaction.

But those are not super strong arguments. Let me know what you think.

galaio replied:
Yeah, `parlia` is not easy import in other low-level modules, it's ok for me to introduce new explicit hooks.

galaio commented on consensus/parlia/parlia.go:
```diff
@@ -1979,7 +1985,36 @@ func (p *Parlia) applyTransaction(
+		if vmConfig.Tracer.OnSystemTxStart != nil {
+			vmConfig.Tracer.OnSystemTxStart()
+		}
+		if tracer.OnTxStart != nil {
```
The `OnTxStart ` may need to be called first, then it can use the right `env`.

maoueh replied:
Sorry not sure I fully understand the comment.

galaio replied:
I mean the `onTxStart` need be called before `OnSystemTxStart`.
```go
if tracer := vmConfig.Tracer; tracer != nil {
		if tracer.OnTxStart != nil {
			tracer.OnTxStart(vmenv.GetVMContext(), expectedTx, msg.From())
		}
		if vmConfig.Tracer.OnSystemTxStart != nil {
			vmConfig.Tracer.OnSystemTxStart()
		}
...
}
```

Because the EVM env only transfers in `OnTxStart`, so it is better to call first. Do I understand correctly? 
```go
func (l *jsonLogger) OnTxStart(env *tracing.VMContext, tx *types.Transaction, from common.Address) {
	l.env = env
}
func (l *jsonLogger) OnSystemTxStart() {
	l.encoder.Encode(l.env.Coinbase, "trigger system tx")
}
```

maoueh replied:
I see the `OnSystemTxStart` acting as a marker. Usually you track the overall state of execution, so I would see an implementation like this:

```
type Tracer struct {
    (... implement Hooks)
    
    inSystemTrx bool
}

func (t *Tracer) OnSystemTrxStart() { t.inSystemTrx = true }
func (t *Tracer) OnTrxStart() { <track trx normally> }
func (t *Tracer) OnTrxEnd() { if (t.inSystemTrx) { // In sytsem } else { // In normal }  }
func (t *Tracer) OnSystemTrxEnd() { t.inSystemTrx = false }
```

In that setup, the system trx start must be called first.

galaio replied:
Ok, but the `defer` order is wrong, may
```go
if vmConfig.Tracer.OnSystemTxEnd != nil {
	defer func() {
		vmConfig.Tracer.OnSystemTxEnd()
	}()
}
if tracer.OnTxEnd != nil {
	defer func() {
		tracer.OnTxEnd(tracingReceipt, applyErr)
	}()
}
```
It will run as you expect:
```go
func (t *Tracer) OnSystemTrxStart() { t.inSystemTrx = true }
func (t *Tracer) OnTrStart() { <track trx normally> }
func (t *Tracer) OnTxEnd() { if (t.inSystemTrx) { // In sytsem } else { // In normal }  }
func (t *Tracer) OnSystemTrxEnd() { t.inSystemTrx = false }
```

maoueh replied:
You are so right, good catch, thanks!

Is there Go tests in the project that exercices the parlia consensus engine and test system transactions that I could tweak to ensure the system tracing is fully covered?

galaio replied:
you can refer `newTestParliaHandlerAfterCancun` in `eth/handler_test.go:256`, there are some integrated testings with parlia. It may help you.

zzzckck commented on consensus/parlia/parlia.go:
```diff
@@ -1979,7 +1985,36 @@ func (p *Parlia) applyTransaction(
+	context := core.NewEVMBlockContext(header, chainContext, nil)
+	// Create a new environment which holds all relevant information
+	// about the transaction and calling mechanisms.
+	vmenv := vm.NewEVM(context, vm.TxContext{Origin: msg.From(), GasPrice: big.NewInt(0)}, state, p.chainConfig, vmConfig)
```
this one change the code logic, the `vmConfig` was empty before, but now it will use the `vmConfig` passed from uplayer.
``` previous
vmenv := vm.NewEVM(context, vm.TxContext{Origin: msg.From(), GasPrice: big.NewInt(0)}, state, chainConfig, vm.Config{})
```
Should be careful.

maoueh replied:
That's true.

I can pass the `tracer` around instead of a pre-built `vm.Config` instance and I could change back to `vm.Config{Tracer: tracer}`. Would you prefer that?

maoueh replied:
@zzzckck

zzzckck replied:
@maoueh I would approve this PR first, I would prefer pass the `tracer` around and change back to `vm.Config{Tracer: tracer}`, could you create another PR for this change?
I would approve and merge this one first, as it is also quite a long time and big PR.

zzzckck reviewed (changes requested):
will hold this PR for the moment, as there would be 2nd round of code sync, which could have conflict.

maoueh commented:
Ok I rebased on latest `develop` and added unit tests, I would have liked to actually test at a higher level, probably using `Blockchain` directly of `StateProcessor`.

But at least the current tests shows overall correct behavior.

maoueh commented:
@zzzckck What's the status on the PR?

zzzckck commented:
> Ok I rebased on latest `develop` and added unit tests, I would have liked to actually test at a higher level, probably using `Blockchain` directly of `StateProcessor`.
> 
> But at least the current tests shows overall correct behavior.

we have almost finished the upstream code sync, we will check if this PR can be included in next release.
